{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecc5e209-bd22-4158-acde-94f82e24d89c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import copy\n",
    "import random\n",
    "from math import log\n",
    "from collections import defaultdict\n",
    "from itertools import islice\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "from scipy import stats\n",
    "from scipy.stats import (\n",
    "    ttest_ind, t, norm, gaussian_kde, pearsonr, f_oneway\n",
    ")\n",
    "from scipy.interpolate import interp1d\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.decomposition import PCA\n",
    "from Bio import Entrez, SeqIO, AlignIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "from Bio.Data import CodonTable\n",
    "from Bio.Phylo.TreeConstruction import DistanceCalculator, DistanceMatrix\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ef8c664-0226-4008-9b70-503c033a7ac5",
   "metadata": {},
   "outputs": [],
   "source": [
    "mit_for_translation = {\"TTT\": \"F\", \"TTC\": \"F\", \"TTA\": \"L\", \"TTG\": \"L\",\n",
    "    \"CTT\": \"L\", \"CTC\": \"L\", \"CTA\": \"L\", \"CTG\": \"L\",\n",
    "    \"ATT\": \"B\", \"ATC\": \"I\", \"ATA\": \"M\", \"ATG\": \"M\",   #ATT is B, provisionally\n",
    "    \"GTT\": \"V\", \"GTC\": \"V\", \"GTA\": \"V\", \"GTG\": \"V\",\n",
    "    \"TCT\": \"S\", \"TCC\": \"S\", \"TCA\": \"S\", \"TCG\": \"S\",\n",
    "    \"CCT\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
    "    \"ACT\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
    "    \"GCT\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
    "    \"TAT\": \"Y\", \"TAC\": \"Y\", \"TAA\": \"*\", \"TAG\": \"*\",\n",
    "    \"CAT\": \"H\", \"CAC\": \"H\", \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
    "    \"AAT\": \"N\", \"AAC\": \"N\", \"AAA\": \"K\", \"AAG\": \"K\",\n",
    "    \"GAT\": \"D\", \"GAC\": \"D\", \"GAA\": \"E\", \"GAG\": \"E\",\n",
    "    \"TGT\": \"C\", \"TGC\": \"C\", \"TGA\": \"W\", \"TGG\": \"W\",\n",
    "    \"CGT\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\",\n",
    "    \"AGT\": \"S\", \"AGC\": \"S\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
    "    \"GGT\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\"}\n",
    "\n",
    "mitochondrial_code_AGRR = {\n",
    "    'A': ['GCT', 'GCC', 'GCA', 'GCG'],\n",
    "    'C': ['TGT', 'TGC'],\n",
    "    'D': ['GAT', 'GAC'],\n",
    "    'E': ['GAA', 'GAG'],\n",
    "    'F': ['TTT', 'TTC'],\n",
    "    'G': ['GGT', 'GGC', 'GGA', 'GGG'],\n",
    "    'H': ['CAT', 'CAC'],\n",
    "    'I': ['ATT', 'ATC'],\n",
    "    'K': ['AAA', 'AAG'],\n",
    "    'L': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
    "    'M': ['ATG', 'ATT', 'ATA'],\n",
    "    'N': ['AAT', 'AAC'],\n",
    "    'P': ['CCT', 'CCC', 'CCA', 'CCG'],\n",
    "    'Q': ['CAA', 'CAG'],\n",
    "    'R': ['CGT', 'CGC', 'CGA', 'CGG'],\n",
    "    'S': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
    "    'T': ['ACT', 'ACC', 'ACA', 'ACG'],\n",
    "    'V': ['GTT', 'GTC', 'GTA', 'GTG'],\n",
    "    'W': ['TGG', 'TGA'],\n",
    "    'Y': ['TAT', 'TAC'],\n",
    "    '*': ['TAA', 'TAG'],\n",
    "    'X': [ 'GCT', 'GCC', 'GCA', 'GCG','TGT', 'TGC','GAT', 'GAC','GAA', 'GAG','TTT', 'TTC','GGT', 'GGC', 'GGA', 'GGG','CAT', 'CAC','ATT', 'ATC','AAA','AAG','TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG','ATG', 'ATT', 'ATA','AAT', 'AAC','CCT', 'CCC', 'CCA', 'CCG','CAA', 'CAG','CGT', 'CGC', 'CGA', 'CGG','TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC','ACT', 'ACC', 'ACA', 'ACG','GTT', 'GTC', 'GTA', 'GTG','TGG', 'TGA','TAT', 'TAC']\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69a89171-d0a7-4c71-a6ea-863f673bf0c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def translate(sequence, genetic_code, for_length = False):\n",
    "    \"\"\"\n",
    "    Translates a nucleotide sequence into an amino acid sequence based on the given genetic code.\n",
    "    \"\"\"\n",
    "    amino_acid_seq = \"\"\n",
    "    # Loop through the sequence in codons (triplets)\n",
    "    for i in range(0, len(sequence), 3):\n",
    "        codon = sequence[i:i+3]  # Extract the codon (triplet)\n",
    "        \n",
    "        if len(codon) == 3:  # Ensure full codon length\n",
    "            amino_acid = genetic_code.get(codon, 'X')  # Translate codon; 'X' for unknown codons\n",
    "            if amino_acid == 'B' and for_length == False:  # we want to keep ATT as B if going to length_DSS\n",
    "                if i == 0:\n",
    "                    amino_acid = 'M'\n",
    "                else: \n",
    "                    amino_acid = 'I'\n",
    "            amino_acid_seq += amino_acid\n",
    "    \n",
    "    return amino_acid_seq\n",
    "\n",
    "def calculate_CpG(sequences, z = 1, local = False):\n",
    "\n",
    "    \"\"\"\n",
    "    sequences = primate sequences\n",
    "    z = number of standard deviations around mean\n",
    "    local means we just want the number of CpG for a single sequence\n",
    "    \n",
    "    returns a mean and a standard deviation * z\n",
    "    \"\"\"\n",
    "    CpG = []\n",
    "    for seq in sequences:\n",
    "        count = sum(1 for i in range(len(seq) - 1) if seq[i:i+2] == 'CG')\n",
    "        CpG.append(count / len(seq))\n",
    "    mu, std = norm.fit(CpG)\n",
    "    if local == True:\n",
    "        return CpG\n",
    "    return mu.round(4), (std*z).round(4)\n",
    "\n",
    "def calculate_GC_content(sequences, z = 1, local = False):\n",
    "    \"\"\"\n",
    "    sequences = primate sequences\n",
    "    z = number of standard deviations around mean\n",
    "    local means we just want the GC content for a single sequence\n",
    "\n",
    "    returns a mean and a standard deviation * z\n",
    "    \"\"\"\n",
    "    GC = []\n",
    "    for seq in sequences:\n",
    "        count = seq.count('G') + seq.count('C')\n",
    "        GC.append(count / len(seq))\n",
    "    mu, std = norm.fit(GC)\n",
    "    if local == True:\n",
    "        #print('GC in function', GC)\n",
    "        return GC\n",
    "    return mu.round(4), (std*z).round(4)\n",
    "\n",
    "def calculate_codon_density(sequences, codons, frame, z):\n",
    "    \"\"\"\n",
    "    sequences = primate sequences\n",
    "    codons is the codon/s for which we want a value\n",
    "    z = number of standard deviations around mean\n",
    "    frame specifies the frame, 0 is reference (+1 in the article), 2 is mtaltco1 (+3 in the article)\n",
    "\n",
    "    returns a mean and a standard deviation * z\n",
    "    \"\"\"\n",
    "    densities = []\n",
    "    for seq in sequences:\n",
    "        shifted_frame = seq[frame:]  # Shift the frame\n",
    "        count = sum(1 for i in range(0, len(shifted_frame) - 2, 3)\n",
    "                    if shifted_frame[i:i+3] in codons)\n",
    "        densities.append(count / (len(seq)//3))\n",
    "\n",
    "    mu, std = norm.fit(densities)\n",
    "\n",
    "    sigma = std*z\n",
    "    \n",
    "    return mu, sigma\n",
    "\n",
    "class phylogenetic_stats:\n",
    "    \"\"\"\n",
    "    gathers all of the listed stats for primates\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, sequences, z=1):\n",
    "        self.sequences = sequences\n",
    "        self.z = z\n",
    "        \n",
    "        self.stop_codons = ['TAG', 'TAA']\n",
    "        self.AGR_codons = ['AGA', 'AGG']\n",
    "        \n",
    "        self.CpG = calculate_CpG(self.sequences, z)\n",
    "        self.GC_content = calculate_GC_content(self.sequences, z)\n",
    "        self.AGR_RF2 = calculate_codon_density(self.sequences, self.AGR_codons, 1, z)\n",
    "        self.AGR_RF3 = calculate_codon_density(self.sequences, self.AGR_codons, 2, z)\n",
    "        self.stop_RF2 = calculate_codon_density(self.sequences, self.stop_codons, 1, z)\n",
    "        self.stop_RF3 = calculate_codon_density(self.sequences, self.stop_codons, 2, z)\n",
    "\n",
    "def plot_two_distributions(mu1, sigma1, mu2, sigma2, label1='', label2=''):\n",
    "    \"\"\"\n",
    "    this is just a visual confirmation we're in the correct frame. We should expect:\n",
    "    high AGR in +3, low AGR in +3, high stop in +2, low stop in +3.\n",
    "    \"\"\"\n",
    "\n",
    "    x_min = min(mu1 - 4*sigma1, mu2 - 4*sigma2)\n",
    "    x_max = max(mu1 + 4*sigma1, mu2 + 4*sigma2)\n",
    "    x = np.linspace(x_min, x_max, 1000)\n",
    "\n",
    "    y1 = norm.pdf(x, mu1, sigma1)\n",
    "    y2 = norm.pdf(x, mu2, sigma2)\n",
    "\n",
    "    plt.figure(figsize=(8, 5))\n",
    "    plt.plot(x, y1, label=label1, color='blue')\n",
    "    plt.plot(x, y2, label=label2, color='red')\n",
    "    plt.fill_between(x, y1, alpha=0.3, color='blue')\n",
    "    plt.fill_between(x, y2, alpha=0.3, color='red')\n",
    "    plt.title('Making sure everything works')\n",
    "    plt.xlabel('Value')\n",
    "    plt.ylabel('Density')\n",
    "    plt.legend()\n",
    "    plt.grid(True)\n",
    "    plt.tight_layout()\n",
    "    plt.show()\n",
    "\n",
    "def local_codon_density(sequence, codons, frame_shift):\n",
    "\n",
    "    \"\"\"\n",
    "    Outputs a given codon density in the given frame_shift of a nucleotide sequence (starts at frame 0)\n",
    "    \"\"\"\n",
    "    \n",
    "    shifted_seq = sequence[frame_shift:]\n",
    "    count = 0\n",
    "    for i in range(0, len(shifted_seq) - 2, 3):\n",
    "        codon = shifted_seq[i:i+3]\n",
    "        if codon in codons:\n",
    "            count += 1\n",
    "    return count / (len(sequence)//3)\n",
    "\n",
    "def generate_DSS(fasta_dict, phylogenetic_stats, fasta_DSS, nb_iter, length = False):\n",
    "\n",
    "    \"\"\"\n",
    "    fasta_dict = every primate nucleotide sequence, in mtCO1's reading frame\n",
    "    phylogenetic stats = sequence constraints\n",
    "    fasta_DSS = dictionary that contains true sequences, unto which this function adds generated sequences (amino acid)\n",
    "    nb_iter = number of generated sequences per item in fasta_dict\n",
    "    length = True or False, if length = True, we maintain ATT as B, if False, ATT is I\n",
    "    \"\"\"\n",
    "    \n",
    "    iteration = 0\n",
    "    label = 'generated'\n",
    "    for i in range(nb_iter):\n",
    "        compte = 0\n",
    "        for key, value in list(fasta_dict.items()): #every primate sequence. value is in nucleotide at this point\n",
    "        \n",
    "            prot_value = translate(value, mit_for_translation)\n",
    "            combinaison, compteur = optimize_sequence(prot_value, mitochondrial_code_AGRR, phylogenetic_stats)\n",
    "            #DSS_seq = translate(combinaison, mit)\n",
    "            fasta_DSS[f\"{label + str(compte)}_{iteration}\"] = translate(combinaison[2:], mit_for_translation, for_length = length)\n",
    "            compte += 1\n",
    "        iteration += 1\n",
    "        print(iteration,'/',nb_iter)\n",
    "    return fasta_DSS\n",
    "\n",
    "def optimize_sequence(protein_sequence, genetic_code, phylogenetic_stats, max_iterations=1000000):\n",
    "\n",
    "    \"\"\"\n",
    "    protein_sequence = CO1 sequence (amino acid)\n",
    "    genetic_code lists every codons that gives a given amino acids\n",
    "    phylogenetic_stats = primate stats for monitored sequence constraints\n",
    "    max_iter, if ever it gets stuck, which it should not (expect around 500 iterations per degenerate sequence)\n",
    "    \"\"\"\n",
    "\n",
    "    stop_codons = phylogenetic_stats.stop_codons\n",
    "    AGR_codons = phylogenetic_stats.AGR_codons\n",
    "\n",
    "    min_stop_2 = phylogenetic_stats.stop_RF2[0] - phylogenetic_stats.stop_RF2[1]\n",
    "    max_stop_2 = phylogenetic_stats.stop_RF2[0] + phylogenetic_stats.stop_RF2[1]\n",
    "    min_AGR_3 = phylogenetic_stats.AGR_RF3[0] - phylogenetic_stats.AGR_RF3[1]\n",
    "    max_AGR_3 = phylogenetic_stats.AGR_RF3[0] + phylogenetic_stats.AGR_RF3[1]\n",
    "    min_GC_content = phylogenetic_stats.GC_content[0] - phylogenetic_stats.GC_content[1]\n",
    "    max_GC_content = phylogenetic_stats.GC_content[0] + phylogenetic_stats.GC_content[1]\n",
    "    min_CpG = phylogenetic_stats.CpG[0] - phylogenetic_stats.CpG[1]\n",
    "    max_CpG = phylogenetic_stats.CpG[0] + phylogenetic_stats.CpG[1]\n",
    "\n",
    "    \n",
    "    current_sequence = ''.join(random.choice(genetic_code[aa]) for aa in protein_sequence)\n",
    "    current_codons = [] \n",
    "    for i in range(0, len(current_sequence), 3):\n",
    "        current_codons.append(current_sequence[i:i+3])\n",
    "        \n",
    "    #print(current_codons)\n",
    "    improved = False\n",
    "    iterations = 0\n",
    "    \n",
    "    def is_better(old_seq, trial_seq):\n",
    "\n",
    "        \"\"\"\n",
    "        compares old and trial seq, if the trial seq is better for any metric, and the old-seq was not within range,\n",
    "        it accepts the trial seq as the new current seq.\n",
    "        \"\"\"\n",
    "\n",
    "        #old\n",
    "        stop2_old = local_codon_density(old_seq, stop_codons, 1)\n",
    "        AGR3_old = local_codon_density(old_seq, AGR_codons, 2)\n",
    "        GC_content_old = calculate_GC_content([old_seq], local = True)\n",
    "        CpG_old = calculate_CpG([old_seq], local = True)\n",
    "\n",
    "        #trial\n",
    "        stop2_trial = local_codon_density(trial_seq, stop_codons, 1)\n",
    "        AGR3_trial = local_codon_density(trial_seq, AGR_codons, 2)\n",
    "        GC_content_trial= calculate_GC_content([trial_seq], local = True)\n",
    "        CpG_trial = calculate_CpG([trial_seq], local = True)\n",
    "\n",
    "        #STOP2 (canonical stops in the +2 frame)\n",
    "        if stop2_old < min_stop_2:\n",
    "            if stop2_trial > stop2_old:\n",
    "                return trial_seq\n",
    "\n",
    "        if stop2_old > max_stop_2:\n",
    "            if stop2_trial < stop2_old:\n",
    "                return trial_seq\n",
    "\n",
    "        #AGR3 (AGR codons in the +3 frame)\n",
    "        if AGR3_old < min_AGR_3:\n",
    "            if AGR3_trial > AGR3_old:\n",
    "                return trial_seq\n",
    "                \n",
    "        if AGR3_old > max_AGR_3:\n",
    "            if AGR3_trial < AGR3_old:\n",
    "                return trial_seq\n",
    "\n",
    "        #GC_content\n",
    "        if GC_content_old < min_GC_content:\n",
    "            if GC_content_trial > GC_content_old:\n",
    "                return trial_seq\n",
    "\n",
    "        if GC_content_old > max_GC_content:\n",
    "            if GC_content_trial < GC_content_old:\n",
    "                return trial_seq\n",
    "        #CpG\n",
    "        if CpG_old < min_CpG:\n",
    "            if CpG_trial > CpG_old:\n",
    "                return trial_seq\n",
    "\n",
    "        if CpG_old > max_CpG:\n",
    "            if CpG_trial < CpG_old:\n",
    "                return trial_seq\n",
    "\n",
    "        #print('not better')\n",
    "        return old_seq\n",
    "                \n",
    "    def is_enough(seq):\n",
    "\n",
    "        \"\"\"\n",
    "        if the current seq meets every criteria, it is good enough and returns True\n",
    "        \"\"\"\n",
    "        \n",
    "        stop2 = local_codon_density(seq, stop_codons, 1)\n",
    "        AGR3 = local_codon_density(seq, AGR_codons, 2)\n",
    "        GC_content = calculate_GC_content([seq], local = True)\n",
    "        CpG = calculate_CpG([seq], local = True)\n",
    "        \n",
    "        return (min_stop_2 <= stop2 <= max_stop_2 and\n",
    "                min_AGR_3 <= AGR3 <= max_AGR_3 and\n",
    "                min_GC_content <= GC_content <= max_GC_content and\n",
    "                min_CpG <= CpG <= max_CpG)\n",
    "\n",
    "    while iterations < max_iterations:\n",
    "        if is_enough(current_sequence):\n",
    "            stop2 = local_codon_density(current_sequence, stop_codons, 1)\n",
    "            AGR3 = local_codon_density(current_sequence, AGR_codons, 2)\n",
    "            GC_content = calculate_GC_content([current_sequence], local = True)\n",
    "            CpG = calculate_CpG([current_sequence], local = True)\n",
    "            return ''.join(current_codons), iterations\n",
    "        \n",
    "        # Mutates one codon at a time\n",
    "        idx = random.randint(0, len(protein_sequence) - 1)\n",
    "        aa = protein_sequence[idx]\n",
    "        alternatives = [c for c in genetic_code[aa] if c != current_codons[idx]] #is it already in use?\n",
    "        \n",
    "        # if no alternative, continue\n",
    "        if not alternatives:\n",
    "            iterations += 1\n",
    "            continue\n",
    "\n",
    "        # else try an alternative codon\n",
    "        trial_codon = random.choice(alternatives)\n",
    "        new_codons = copy.deepcopy(current_codons)\n",
    "        new_codons[idx] = trial_codon\n",
    "        trial_seq = ''.join(new_codons)\n",
    "        \n",
    "        current_sequence = is_better(trial_seq, current_sequence)\n",
    "        current_codons = []\n",
    "        \n",
    "        for i in range(0, len(current_sequence), 3):\n",
    "            current_codons.append(current_sequence[i:i+3])\n",
    "        iterations += 1\n",
    "        if iterations % 1000 == 0:\n",
    "            print(f\"Iteration: {iterations}\")\n",
    "\n",
    "    return ''.join(current_codons), iterations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4ea2d65-bcc3-4668-9884-d7f06d8d76ed",
   "metadata": {},
   "source": [
    "# Import 27 primate sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86448de2-d8f9-41e2-a5b5-ecc36fb00e48",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_dict = {} # nucleotide-level\n",
    "\n",
    "for record in SeqIO.parse(\"../data/27_primates_mtaltCO1_nuc.fasta\", \"fasta\"):\n",
    "    fasta_dict[record.description] = str(record.seq[1:-1]) #to be in the CO1 frame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3166753-6258-4aef-ba6e-52137fdcc405",
   "metadata": {},
   "source": [
    "# Compile primate sequence constraints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22595e0b-a919-48e8-9770-d00fe87541f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_DSS = {} # protein level, +3 RF\n",
    "\n",
    "listes = []\n",
    "label = 'true'\n",
    "compteur = 0\n",
    "for key, value in list(fasta_dict.items()):\n",
    "    fasta_DSS[f\"{label + str(compteur)}\"] = translate(value[2:], mit_for_translation) #skip first two \n",
    "    listes.append(value)\n",
    "    compteur += 1\n",
    "\n",
    "primate_DSS = fasta_DSS\n",
    "length_DSS = {}\n",
    "\n",
    "stats_primates = phylogenetic_stats(listes, z = 1)\n",
    "\n",
    "plot_two_distributions(stats_primates.AGR_RF2[0],stats_primates.AGR_RF2[1],stats_primates.AGR_RF3[0], stats_primates.AGR_RF3[1], label1 = 'AGR_+2', label2 = 'AGR_+3')\n",
    "plot_two_distributions(stats_primates.stop_RF2[0],stats_primates.stop_RF2[1],stats_primates.stop_RF3[0], stats_primates.stop_RF3[1], label1 = 'stop_+2', label2 = 'stop_+3')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "638b45e8-f284-40b7-9f6b-b1105e868a17",
   "metadata": {},
   "source": [
    "# Generate degenerate sequences for primate PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48752b20-ac44-47a8-b57a-d072c5ee0154",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "output_DSS = generate_DSS(fasta_dict, stats_primates, primate_DSS,20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6348d67e-16a0-4ab0-812a-0f0dcd0852d8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "file_path = \"../data/primate_DSS.fasta\"\n",
    "\n",
    "with open(file_path, \"w\") as f:\n",
    "    for key, value in output_DSS.items():\n",
    "        print(key,len(value))\n",
    "        if 'X' not in value: #for benner74\n",
    "            header = f\"> {key}\\n\"\n",
    "            sequence = value.replace(\"*\", \"*\") + \"\\n\"\n",
    "            f.write(header + sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4ef675f-21c5-496e-a786-7ba091ffcd27",
   "metadata": {},
   "source": [
    "# Generate homo sapiens degenerate sequences for length comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89bfed25-ed7b-4dd2-b230-4da0730d004e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    " # just make degenerate sequences from Homo sapiens' CO1 sequence (first in fasta_dict)\n",
    "first = dict([next(iter(fasta_dict.items()))])\n",
    "\n",
    "length_DSS = {}\n",
    "length_DSS = generate_DSS(first, stats_primates, length_DSS,1000, length = True) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c9cf8e6-7505-4de2-b602-16b2f7c2f19b",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = \"../data/length_DSS.fasta\"\n",
    "\n",
    "with open(file_path, \"w\") as f:\n",
    "    for key, value in length_DSS.items():\n",
    "        #print(key,len(value))\n",
    "        if 'X' not in value: #for benner74\n",
    "            header = f\"> {key}\\n\"\n",
    "            sequence = value.replace(\"*\", \"*\") + \"\\n\"\n",
    "            f.write(header + sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "735a9a6b-d867-4f90-8503-af9a84fdbf08",
   "metadata": {},
   "source": [
    "# Generate homo sapiens degenerate sequence for disorder predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7db3bde0-ee08-44dd-a1a4-ce0194daa574",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = \"../data/disorder_DSS.fasta\" \n",
    "\n",
    "# Ridao only handles 100 unknown samples at a time\n",
    "max_sample = 100\n",
    "sample  = 0\n",
    "\n",
    "with open(file_path, \"w\") as f:\n",
    "    for key, value in length_DSS.items():\n",
    "        if sample < max_sample:\n",
    "            #print(key,len(value))\n",
    "            if 'X' not in value: #for benner74\n",
    "                header = f\"> {key}\\n\"\n",
    "                sequence = value.replace(\"*\", \"\") + \"\\n\" # Ridao does not handle *\n",
    "                sequence = value.replace(\"B\", \"I\") + \"\\n\" \n",
    "                f.write(header + sequence)\n",
    "        sample += 1\n",
    "        "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
