{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Code pr√©servation"
      ],
      "metadata": {
        "id": "Lb-TnFOOK5YH"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "rBFW3w3HK2Z9"
      },
      "outputs": [],
      "source": [
        "pip install Bio"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import re\n",
        "import csv\n",
        "from Bio import SeqIO\n",
        "from Bio.Seq import Seq\n",
        "\n",
        "#Genetic code, mitochondrial or mtalt\n",
        "mitalt_genetic_code = {\n",
        "    \"TTT\" : \"F\", \"TTC\" : \"F\", \"TTA\" : \"L\", \"TTG\" : \"L\",\n",
        "    \"TCT\" : \"S\", \"TCC\" : \"S\", \"TCA\" : \"S\", \"TCG\" : \"S\",\n",
        "    \"TAT\" : \"Y\", \"TAC\" : \"Y\", \"TAA\" : \"*\", \"TAG\" : \"*\",\n",
        "    \"TGT\" : \"C\", \"TGC\" : \"C\", \"TGA\" : \"W\", \"TGG\" : \"W\",\n",
        "    \"CTT\" : \"L\", \"CTC\" : \"L\", \"CTA\" : \"L\", \"CTG\" : \"L\",\n",
        "    \"CCT\" : \"P\", \"CCC\" : \"P\", \"CCA\" : \"P\", \"CCG\" : \"P\",\n",
        "    \"CAT\" : \"H\", \"CAC\" : \"H\", \"CAA\" : \"Q\", \"CAG\" : \"Q\",\n",
        "    \"CGT\" : \"R\", \"CGC\" : \"R\", \"CGA\" : \"R\", \"CGG\" : \"R\",\n",
        "    \"ATT\" : \"I\", \"ATC\" : \"I\", \"ATA\" : \"I\", \"ATG\" : \"M\",\n",
        "    \"ACT\" : \"T\", \"ACC\" : \"T\", \"ACA\" : \"T\", \"ACG\" : \"T\",\n",
        "    \"AAT\" : \"N\", \"AAC\" : \"N\", \"AAA\" : \"K\", \"AAG\" : \"K\",\n",
        "    \"AGT\" : \"S\", \"AGC\" : \"S\", \"AGA\" : \"R\", \"AGG\" : \"R\",\n",
        "    \"GTT\" : \"V\", \"GTC\" : \"V\", \"GTA\" : \"V\", \"GTG\" : \"V\",\n",
        "    \"GCT\" : \"A\", \"GCC\" : \"A\", \"GCA\" : \"A\", \"GCG\" : \"A\",\n",
        "    \"GAT\" : \"D\", \"GAC\" : \"D\", \"GAA\" : \"E\", \"GAG\" : \"E\",\n",
        "    \"GGT\" : \"G\", \"GGC\" : \"G\", \"GGA\" : \"G\", \"GGG\" : \"G\"\n",
        " }\n",
        "\n",
        "nuclear_genetic_code = {\n",
        "    \"TTT\" : \"F\", \"TTC\" : \"F\", \"TTA\" : \"L\", \"TTG\" : \"L\",\n",
        "    \"TCT\" : \"S\", \"TCC\" : \"S\", \"TCA\" : \"S\", \"TCG\" : \"S\",\n",
        "    \"TAT\" : \"Y\", \"TAC\" : \"Y\", \"TAA\" : \"*\", \"TAG\" : \"*\",\n",
        "    \"TGT\" : \"C\", \"TGC\" : \"C\", \"TGA\" : \"*\", \"TGG\" : \"W\",\n",
        "    \"CTT\" : \"L\", \"CTC\" : \"L\", \"CTA\" : \"L\", \"CTG\" : \"L\",\n",
        "    \"CCT\" : \"P\", \"CCC\" : \"P\", \"CCA\" : \"P\", \"CCG\" : \"P\",\n",
        "    \"CAT\" : \"H\", \"CAC\" : \"H\", \"CAA\" : \"Q\", \"CAG\" : \"Q\",\n",
        "    \"CGT\" : \"R\", \"CGC\" : \"R\", \"CGA\" : \"R\", \"CGG\" : \"R\",\n",
        "    \"ATT\" : \"I\", \"ATC\" : \"I\", \"ATA\" : \"I\", \"ATG\" : \"M\",\n",
        "    \"ACT\" : \"T\", \"ACC\" : \"T\", \"ACA\" : \"T\", \"ACG\" : \"T\",\n",
        "    \"AAT\" : \"N\", \"AAC\" : \"N\", \"AAA\" : \"K\", \"AAG\" : \"K\",\n",
        "    \"AGT\" : \"S\", \"AGC\" : \"S\", \"AGA\" : \"R\", \"AGG\" : \"R\",\n",
        "    \"GTT\" : \"V\", \"GTC\" : \"V\", \"GTA\" : \"V\", \"GTG\" : \"V\",\n",
        "    \"GCT\" : \"A\", \"GCC\" : \"A\", \"GCA\" : \"A\", \"GCG\" : \"A\",\n",
        "    \"GAT\" : \"D\", \"GAC\" : \"D\", \"GAA\" : \"E\", \"GAG\" : \"E\",\n",
        "    \"GGT\" : \"G\", \"GGC\" : \"G\", \"GGA\" : \"G\", \"GGG\" : \"G\"\n",
        "    }\n",
        "\n",
        "mitochondrial_genetic_code = {\n",
        "    \"TTT\" : \"F\", \"TTC\" : \"F\", \"TTA\" : \"L\", \"TTG\" : \"L\",\n",
        "    \"TCT\" : \"S\", \"TCC\" : \"S\", \"TCA\" : \"S\", \"TCG\" : \"S\",\n",
        "    \"TAT\" : \"Y\", \"TAC\" : \"Y\", \"TAA\" : \"*\", \"TAG\" : \"*\",\n",
        "    \"TGT\" : \"C\", \"TGC\" : \"C\", \"TGA\" : \"W\", \"TGG\" : \"W\",\n",
        "    \"CTT\" : \"L\", \"CTC\" : \"L\", \"CTA\" : \"L\", \"CTG\" : \"L\",\n",
        "    \"CCT\" : \"P\", \"CCC\" : \"P\", \"CCA\" : \"P\", \"CCG\" : \"P\",\n",
        "    \"CAT\" : \"H\", \"CAC\" : \"H\", \"CAA\" : \"Q\", \"CAG\" : \"Q\",\n",
        "    \"CGT\" : \"R\", \"CGC\" : \"R\", \"CGA\" : \"R\", \"CGG\" : \"R\",\n",
        "    \"ATT\" : \"I\", \"ATC\" : \"I\", \"ATA\" : \"I\", \"ATG\" : \"M\",\n",
        "    \"ACT\" : \"T\", \"ACC\" : \"T\", \"ACA\" : \"T\", \"ACG\" : \"T\",\n",
        "    \"AAT\" : \"N\", \"AAC\" : \"N\", \"AAA\" : \"K\", \"AAG\" : \"K\",\n",
        "    \"AGT\" : \"S\", \"AGC\" : \"S\", \"AGA\" : \"*\", \"AGG\" : \"*\",\n",
        "    \"GTT\" : \"V\", \"GTC\" : \"V\", \"GTA\" : \"V\", \"GTG\" : \"V\",\n",
        "    \"GCT\" : \"A\", \"GCC\" : \"A\", \"GCA\" : \"A\", \"GCG\" : \"A\",\n",
        "    \"GAT\" : \"D\", \"GAC\" : \"D\", \"GAA\" : \"E\", \"GAG\" : \"E\",\n",
        "    \"GGT\" : \"G\", \"GGC\" : \"G\", \"GGA\" : \"G\", \"GGG\" : \"G\"\n",
        " }\n",
        "\n",
        "#Store variants\n",
        "class Var:\n",
        "    def __init__(self, index, locus, position, init, final, overall_frequency, frequency_L, frequency_M, frequency_N):\n",
        "        self.index = int(index)\n",
        "        self.locus = locus\n",
        "        self.position = int(position)\n",
        "        self.init = init\n",
        "        self.final = final\n",
        "        self.overall_frequency = float(overall_frequency)\n",
        "        self.frequency_L = float(frequency_L)\n",
        "        self.frequency_M = float(frequency_M)\n",
        "        self.frequency_N = float(frequency_N)\n",
        "\n",
        "#Store ORF\n",
        "class ORF:\n",
        "    def __init__(self, sequence, start, end, expected_protein, genetic_code, name):\n",
        "        self.sequence_ADN_initiale = sequence\n",
        "        self.sequence_ADN = sequence\n",
        "        self.start = start\n",
        "        self.end = end\n",
        "        self.variants = []\n",
        "        self.genetic_code = genetic_code\n",
        "        self.preservation_overall = 100\n",
        "        self.preservation_N = 100\n",
        "        self.preservation_M = 100\n",
        "        self.preservation_L = 100\n",
        "        self.expected_protein = expected_protein\n",
        "        self.length = len(expected_protein) #Initial protein length\n",
        "        self.name = name\n",
        "        self.destructive_variants = []  #Destructive variants\n",
        "\n",
        "    def translate_sequence(self, sequence, strict=False):\n",
        "        protein = \"\"\n",
        "        start = 0\n",
        "\n",
        "        if not strict:\n",
        "            for i in range(0, len(sequence) - 2, 3):\n",
        "                codon = sequence[i:i+3]\n",
        "                if codon in [\"ATG\", \"ATA\", \"ATT\"]:\n",
        "                    start = i\n",
        "                    break\n",
        "\n",
        "        first = True\n",
        "        for i in range(start, len(sequence) - 2, 3):\n",
        "            codon = sequence[i:i+3]\n",
        "\n",
        "            # Check if it's the first codon\n",
        "            if first:\n",
        "                if codon in [\"ATG\", \"ATA\", \"ATT\"]:\n",
        "                    aa = \"M\"\n",
        "                else:\n",
        "                    aa = self.genetic_code.get(codon, \"?\")\n",
        "                first = False\n",
        "            else:\n",
        "                aa = self.genetic_code.get(codon, \"?\")\n",
        "\n",
        "            protein += aa\n",
        "            if aa == \"*\":\n",
        "                break\n",
        "\n",
        "        return protein\n",
        "\n",
        "\n",
        "    def apply_variants(self, variants_list):\n",
        "        sequence_list = list(self.sequence_ADN_initiale)  # Start from intial protein\n",
        "        initial_protein = self.expected_protein\n",
        "        initial_length = self.length\n",
        "\n",
        "        self.variants = []\n",
        "        self.destructive_variants = []\n",
        "\n",
        "        for var in variants_list:\n",
        "          if not is_variant_applicable_to_orf(self, var):\n",
        "              continue\n",
        "\n",
        "          variant_pos = var.position - self.start\n",
        "          sequence_list[variant_pos] = var.final\n",
        "          self.variants.append(var)\n",
        "\n",
        "          # Is the variant destructive?\n",
        "          temp_seq = list(self.sequence_ADN_initiale)\n",
        "          temp_seq[variant_pos] = var.final\n",
        "          mutated_temp_protein = self.translate_sequence(\"\".join(temp_seq), strict=False)\n",
        "\n",
        "          if len(mutated_temp_protein) != initial_length:\n",
        "              self.destructive_variants.append(var)\n",
        "              self.preservation_overall -= var.overall_frequency\n",
        "              self.preservation_L -= var.frequency_L\n",
        "              self.preservation_M -= var.frequency_M\n",
        "              self.preservation_N -= var.frequency_N\n",
        "\n",
        "\n",
        "        # Update mutated protein\n",
        "        self.sequence_ADN = \"\".join(sequence_list)\n",
        "        mutated_protein = self.translate_sequence(self.sequence_ADN, strict=False)\n",
        "\n",
        "        return mutated_protein\n",
        "\n",
        "def is_variant_applicable_to_orf(orf, var):\n",
        "    relative_pos = var.position - orf.start\n",
        "    if not (orf.start <= var.position <= orf.end):\n",
        "        return False\n",
        "    if not (0 <= relative_pos < len(orf.sequence_ADN_initiale)):\n",
        "        return False\n",
        "    if orf.sequence_ADN_initiale[relative_pos] != var.init:\n",
        "        return False\n",
        "    return True\n",
        "\n",
        "# Mitochondrial genome\n",
        "def get_mitochondrial_dna(filename=\"/content/human_mitochondrial_genome.fasta\"):\n",
        "    with open(filename) as handle:\n",
        "        record = SeqIO.read(handle, \"fasta\")\n",
        "        return record.seq\n",
        "\n",
        "# Read fasta with ORFs\n",
        "def read_fasta_proteins(fasta_file, genome_seq, genetic_code):\n",
        "    orfs = []\n",
        "    with open(fasta_file) as handle:\n",
        "        for record in SeqIO.parse(handle, \"fasta\"):\n",
        "            header = record.description\n",
        "            match = re.search(r\"start: (\\d+) end: (\\d+)\", header)\n",
        "            if match:\n",
        "                start, end = map(int, match.groups())\n",
        "                expected_protein = str(record.seq)\n",
        "                dna_seq = genome_seq[start:end]\n",
        "\n",
        "                orf = ORF(dna_seq, start, end, expected_protein, 3, genetic_code, record.id)\n",
        "                orfs.append(orf)\n",
        "\n",
        "    return orfs\n",
        "\n",
        "# Variants\n",
        "def read_variants_file(variant_file):\n",
        "    variants_list = []\n",
        "    with open(variant_file, newline='') as csvfile:\n",
        "        variants = csv.reader(csvfile, delimiter=';', quotechar='\"')\n",
        "        next(variants)\n",
        "        for row in variants:\n",
        "            index, locus, position, init, final, overall_frequency, frequency_L, frequency_M, frequency_N = row\n",
        "            variants_list.append(Var(index, locus, position, init, final, overall_frequency, frequency_L, frequency_M, frequency_N))\n",
        "    return variants_list\n",
        "\n",
        "# Save in a csv\n",
        "def save_results_to_csv(orfs, filename=\"orf_analysis_results_mtaltco1_overall.csv\"):\n",
        "    with open(filename, mode=\"w\", newline=\"\") as file:\n",
        "        writer = csv.writer(file, delimiter=\";\")\n",
        "        writer.writerow([\"Nom ORF\", \"Start\", \"End\", \"ADN initial\",\n",
        "                         \"Prot√©ine initiale\", \"Variants appliqu√©s\",\n",
        "                         \"Variants destructeurs\", \"Preservation Score Overall\", \"Preservation Score Haplogroup L\", \"Preservation Score Haplogroup M\", \"Preservation Score Haplogroup N\"])\n",
        "\n",
        "        for orf in orfs:\n",
        "            mutated_protein = orf.translate_sequence(orf.sequence_ADN, strict=True)\n",
        "            applied_variants = \";\".join([f\"{v.position}:{v.init}>{v.final} ({v.overall_frequency}, {v.frequency_L}, {v.frequency_M}, {v.frequency_N})\" for v in orf.variants])\n",
        "            destructive_variants = \";\".join([f\"{v.position}:{v.init}>{v.final} ({v.overall_frequency}, {v.frequency_L}, {v.frequency_M}, {v.frequency_N})\" for v in orf.destructive_variants])\n",
        "            writer.writerow([orf.name, orf.start, orf.end, orf.sequence_ADN_initiale,\n",
        "                             orf.expected_protein, applied_variants, destructive_variants, f\"{orf.preservation_overall:.6f}\", f\"{orf.preservation_L:.6f}\", f\"{orf.preservation_M:.6f}\", f\"{orf.preservation_N:.6f}\"])\n",
        "\n",
        "# Save the log\n",
        "def save_log(orfs, variants_list, log_filename=\"orf_analysis_log_overall.txt\"):\n",
        "    with open(log_filename, \"w\") as log_file:\n",
        "        for orf in orfs:\n",
        "            log_file.write(f\"===== ORF: {orf.name} (start: {orf.start}, end: {orf.end}) =====\\n\")\n",
        "            initial_protein = orf.translate_sequence(orf.sequence_ADN_initiale)\n",
        "            sequence_list = list(orf.sequence_ADN_initiale)\n",
        "\n",
        "            for var in variants_list:\n",
        "                log_file.write(f\"\\nVariant {var.index}, position : {var.position}: {var.init} > {var.final} - L : {var.frequency_L} - M : {var.frequency_M} - N : {var.frequency_N} \\n\")\n",
        "\n",
        "                # Step 1: Is it in the range?\n",
        "                if not (orf.start <= var.position <= orf.end):\n",
        "                    log_file.write(\"En dehors du range de cette ORF.\\n\")\n",
        "                    continue\n",
        "\n",
        "                relative_pos = var.position - orf.start\n",
        "\n",
        "                # Relative position in range\n",
        "                if relative_pos < 0 or relative_pos >= len(sequence_list):\n",
        "                    log_file.write(f\"Position relative hors des limites de l‚ÄôORF (relative_pos = {relative_pos}).\\n\")\n",
        "                    continue\n",
        "\n",
        "                # Step 2: Does the initial nucleotide match the variant one?\n",
        "                if sequence_list[relative_pos] != var.init:\n",
        "                    log_file.write(f\"Nucl√©otide attendu ({var.init}) ‚â† trouv√© ({sequence_list[relative_pos]}). Mutation non appliqu√©e.\\n\")\n",
        "                    continue\n",
        "\n",
        "                log_file.write(\"Dans le range et nucl√©otide initial correct. Application de la mutation.\\n\")\n",
        "\n",
        "                # Step 3: Apply the mutation\n",
        "                temp_sequence = sequence_list.copy()\n",
        "                temp_sequence[relative_pos] = var.final\n",
        "                mutated_seq = \"\".join(temp_sequence)\n",
        "\n",
        "                # Step 4: Translate and compare\n",
        "                mutated_protein = orf.translate_sequence(mutated_seq)\n",
        "\n",
        "                if len(mutated_protein) != len(initial_protein):\n",
        "                    log_file.write(f\"Prot√©ine mut√©e diff√©rente en longueur !\\n\")\n",
        "                    log_file.write(f\"Prot√©ine mut√©e :\\n{mutated_protein}\\n\")\n",
        "                else:\n",
        "                    log_file.write(\"Prot√©ine mut√©e de m√™me longueur que l'initiale.\\n\")\n",
        "\n",
        "            log_file.write(\"\\n\\n\")\n",
        "\n",
        "\n",
        "# ORFs\n",
        "my_orf = ORF(\n",
        "    sequence=\"ATTTGTAATAATCTTCTTCATAGTAATACCCATCATAATCGGAGGCTTTGGCAACTGACTAGTTCCCCTAATAATCGGTGCCCCCGATATGGCGTTTCCCCGCATAAACAACATAAGCTTCTGACTCTTACCTCCCTCTCTCCTACTCCTGCTCGCATCTGCTATAGTGGAGGCCGGAGCAGGAACAGGTTGAACAGTCTACCCTCCCTTAGCAGGGAACTACTCCCACCCTGGAGCCTCCGTAGACCTAACCATCTTCTCCTTACACCTAGCAGGTGTCTCCTCTATCTTAGGGGCCATCAATTTCATCACAACAATTATCAATATAAAACCCCCTGCCATAACCCAATACCAAACGCCCCTCTTCGTCTGATCCGTCCTAATCACAGCAGTCCTACTTCTCCTATCTCTCCCAGTCCTAGCTGCTGGCATCACTATACTACTAACAGACCGCAACCTCAACACCACCTTCTTCGACCCCGCCGGAGGAGGAGACCCCATTCTATACCAACACCTATTCTGATTTTTCGGTCACCCTGAAGTTTATATTCTTATCCTACCAGGCTTCGGAATAATCTCCCATATTGTAACTTACTACTCCGGAAAAAAAGAACCATTTGGATACATAGGTATGGTCTGAGCTATGATATCAATTGGCTTCCTAGGGTTTATCGTGTGAGCACACCATATATTTACAGTAGGAATAGACGTAGACACACGAGCATATTTCACCTCCGCTACCATAATCATCGCTATCCCCACCGGCGTCAAAGTATTTAG\",  # Par exemple, ORF de 0 √† 90\n",
        "    start=6089,\n",
        "    end=6868, #6866 and the 3 nt from the stop codon\n",
        "    expected_protein=\"MCNNLLHSNTHHNRRLWQLTSSPNNRCPRYGVSPHKQHKLLTLTSLSPTPARICYSGGRSRNRLNSLPSLSRELLPPWSLRRPNHLLLTPSRCLLYLRGHQFHHNNYQYKTPCHNPMPNAPLRLIRPNHSSPTSPISPSPSCWHHYTTNRPQPQHHLLRPRRRRRPHSMPTPILIFRSPWSLYSYPTRLRNNLPYCNLLLRKKRTIWMHRYGLSYDINWLPRVYRVSTPYIYSRNRRRHTSMFHLRYHNHRYPHRRQSI*\",   # Ou la prot√©ine attendue si tu l‚Äôas\n",
        "    genetic_code=mitalt_genetic_code,\n",
        "    name=\"MTALTCO1\"\n",
        ")\n",
        "\n",
        "orfs = [my_orf]\n",
        "\n",
        "#orfs = read_fasta_proteins(\"MTALTCO1.fasta\", genome_seq, mitalt_genetic_code)# or mitochondrial_genetic_code\n",
        "\n",
        "# Variants\n",
        "variants_list = read_variants_file(\"variants.csv\")\n",
        "\n",
        "# Apply variants\n",
        "for orf in orfs:\n",
        "    orf.apply_variants(variants_list)\n",
        "\n",
        "# Save results\n",
        "save_results_to_csv(orfs)\n",
        "save_log(orfs, variants_list)\n",
        "\n",
        "\n",
        "print(\"R√©sultats sauvegard√©s dans orf_analysis_results_overall.csv\")\n",
        "\n",
        "print(\"Log d√©taill√© sauvegard√© dans orf_analysis_log_overall.txt\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6Yt-OWbeQ3wO",
        "outputId": "9a9e7b9d-7458-4128-e4d3-0ee00662da4e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "R√©sultats sauvegard√©s dans orf_analysis_results_overall.csv\n",
            "Log d√©taill√© sauvegard√© dans orf_analysis_log_overall.txt\n"
          ]
        }
      ]
    }
  ]
}